### PR TITLE
floating save button

### LINKED ISSUE #1525: Make saving easier (e.g. with a floating save button)
This will mitigate some of the problems discussed in #979

### PR DESCRIPTION
**Pull request by opqdonut**
closes #1525

### CODE REVIEW

**Comment by opqdonut** [comment]
The two commits here present two options for a floating actions div. Which do you prefer?

## float bottom
float-bottom

## float right
float-right

**Comment by hukka** [comment]
Float on right stands out more, and leaves more space for the actual form. Can we do both easily, with breakpoints, so that we use the horizontal space when it's available?

**Comment by opqdonut** [comment]
We looked at this with @Macroz and moved the actions pane below the application title

Responsiveness works: in desktop & tablet sizes we get the floating div. In mobile sizes the actions stay at the bottom.

The remaining issue is that saving a draft reloads the application and scrolls to the top, which kinda defeats the purpose of a floating button. Waiting for #1469 to get implemented first.

**Comment by opqdonut** [comment]
Rebased on top of fresh master.

**Comment by luontola** [comment]
#1549 has been merged, so this can proceed.

**Comment by opqdonut** [comment]
Flash messages now appear above the floating actions div and don't scroll the page.

As a side-effect, in phone layout the actions div is at the top.

**Comment by opqdonut** [comment]
I think this is ready now

**Comment by okahilak** [comment]
Hmm, so now it looks like this:

image

I might prefer the message element to be below the action panel, that somehow corresponds more to the causal direction of the action and its message.

File: src/cljs/rems/focus.cljs [hunk 1]
```diff
@@ -2,13 +2,18 @@
...
-         (.setAttribute element "tabindex" "-1")
-         (.focus element))
-       (js/setTimeout #(focus-element-async selector (dec tries))
-                      10)))))
+  "Focus an element when it appears. Options can include:
+    :tries -- number of times to poll, defaults to 100
+    :scroll? -- whether to .scollIntoView the element"
+  [selector & [options]]
+  (let [tries (get options :tries 100)
+        scoll? (get options :scroll? false)]
```
Comment (okahilak) [inline src/cljs/rems/focus.cljs]: There's a typo here. It seems to work anyway, which is kind of funny & strange. However, I assume that having :scroll? as true wouldn't work.

File: src/cljs/rems/focus.cljs [hunk 2]
```diff
@@ -2,13 +2,18 @@
...
-   (when (pos? tries)
-     (if-let [element (.querySelector js/document selector)]
-       (do
-         (.setAttribute element "tabindex" "-1")
-         (.focus element))
-       (js/setTimeout #(focus-element-async selector (dec tries))
-                      10)))))
+  "Focus an element when it appears. Options can include:
+    :tries -- number of times to poll, defaults to 100
+    :scroll? -- whether to .scollIntoView the element"
```
Comment (okahilak) [inline src/cljs/rems/focus.cljs]: a typo

**Comment by okahilak** [comment]
> I might prefer the message element to be below the action panel, that somehow corresponds more to the causal direction of the action and its effects.

Ok, so this might be difficult to get to work together with the alternative position for the action panel on the top of the page, in which case it is nice to have the flash message element above the panel.

So, all in all, I'm not proposing this change in this patch after all. I think we can fine-tune the positioning later on, if need be.

File: src/cljs/rems/flash_message.cljs [hunk 3]
```diff
@@ -19,8 +19,7 @@
    ;; On Chrome, focusing the element scrolls it fully into view,
    ;; but on Firefox the element is hidden behind the navigation menu,
    ;; so explicit scrolling is needed.
```
Comment (opqdonut) [inline src/cljs/rems/flash_message.cljs]: this bug now happens again in responsive mode. I'll have a look if I can fix it...

Comment (opqdonut) [reply src/cljs/rems/flash_message.cljs]: created #1578

Comment (opqdonut) [reply src/cljs/rems/flash_message.cljs]: created #1578

**Comment by foxlynx** [comment]
Needs a bit more consistency in success announcements:

* better feedback for adding attachment
* better feedback when inviting/adding member

Now the two cases above appear to the UI without notification or highlight colors, which is ok if it is consistent I guess, but when accepting terms there is a green higlighted success box on spot, so it is confusing for user

**Comment by opqdonut** [comment]
Some stuff that came up while working on this: #1580 #1581

**Pull request merged**